README.md

```diff
@@ -1,42 +1,143 @@
+While building the image:
+
+* Installs postgres majors' `pg_config`. We only need `pg_config` to generate citus artifacts.
+
+While running the container:
+
+* Installs and tars the checked out citus version for the postgres versions. This is done so that the other jobs can install citus easily by untarring without the need to do `make install`. We need to do this step in the running container, not while building the image, so that the built citus version is the checked out citus code.
+* Creates `build-{pg_version}` folders with citus configured, so these folders have the necessary `Makefile.global` to run the tests. These folders are generated for each postgres version. For example, at the time of writing there are 3 folders with `build-10`, `build-11` and `build-12`.
 
 The general process for other images are similar, which is:
 
-- Untar the citus tar to install citus for the postgres version that the image will use. For example, if postgres 11 is used, the tar is named `install-11.tar`.
-- The script (`install-and-test-ext`) takes the target name as an argument, and the following command is used to run the test:
+While building the image:
+
+* Install postgres major which is taken as an argument `PG_MAJOR`.
+
+While running the container:
+
+* Untar the citus tar to install citus for the postgres version that the image will use. For example, if postgres 11 is used, the tar is named `install-11.tar`.
+* The script (`install-and-test-ext`) takes the target name as an argument, and the following command is used to run the test:
 
 ```bash
 gosu circleci make -C "${CIRCLE_WORKING_DIRECTORY}/build-${PG_MAJOR}/src/test/regress" "${@}"
 ```
 
-- Finally `regression.diffs` is printed if it exists, which indicates that there was a problem during the tests.
+* Finally `regression.diffs` is printed if it exists, which indicates that there was a problem during the tests.
+
+Because of the specific logic that we have to run tests faster, we have a docker image `debbuilder` that generates a debian package for `postgresql-server-dev-{pg-major}`. The generated package is in the built docker image, specifically in `{container-name}/home/circleci/debs`. The package is pushed to [the-process](https://packagecloud.io/citus-bot/the-process). This package place is added to apt source list and pinned with with the highest priority so that it is chosen over the standard `postgresql-server-dev-{pg-major}` package.
+
+The installation of postgres packages is at the time of building the image to reduce testing time. However this means that once the image is built the postgres version is fixed. You will need to rebuild the image if you want to upgrade the image, upgrading the image for a minor version is easier compared to upgrading a major version.
```

Comment:
shouldn't changing the major version for most containers entail to rebuild with a different `PG_MAJOR` build arg? I think only for pg upgrade tests, where we need multiple postgres versions, and the extbuilder, where we also need multiple postgres versions, the build is slightly harder as we need to change some of the scripts.

Reply:
For many images yes.  What should be done is explained for each image, is there something that you think is redundant or can be simplified?